### ISSUE TITLE
Cannot override compilers inside an environment

### ISSUE THREAD

**Issue by scottwittenburg** [spack/spack#13442]
TLDR: `with cfg.override('compilers:', actual_compilers):` does not seem actually to override the compilers inside a spack environment.

When running pipelines that are intended to bootstrap a missing compiler, we have typically added the compiler we want to the spack environment, substituting empty strings for all the paths.  This allows us to "cross-concretize" the specs against that compiler even if the system where we concretize doesn't actually have that compiler.

Now we are attempting to build that compiler and it's dependencies within the environment where we have listed it only for cross-concretization purposes, but it's getting picked as the compiler to build with.  For example, say my release environment wants to build:

```
bzip2@1.0.8@gcc@5.5.0 os=centos7
```

But I'm sure the machine where I want to build that doesn't have `gcc@5.5.0`, so I put it in my environment like this:

```
compilers:
    - compiler:
        spec: gcc@5.5.0
        target: x86_64
        operating_system: centos7
        paths: { "cc": "", "cxx": "", "f77": "", "fc": "" }
        environment: {}
        extra_rpaths: []
        flags: {}
        modules: []
```

And that allows me to concretize `bzip2` with the right compiler.  Then I'll add `gcc@5.5.0` as a compiler to bootstrapped in that environment (within the `gitlab-ci` section).  When I'm concretizing `gcc@5.5.0` to be able to stage it and all it's dependencies in the bootstrapping portion of the pipeline, I want spack to use whatever compiler is available on that machine to build `gcc@5.5.0`, so I try to explicitly override the compilers configuration to remove the compiler I know I only put there for cross-concretization, so that spack chooses an actual compiler on the system with which to build `gcc@5.5.0`.  This is happening in python and looks something like:

```python
import spack.config as cfg
from spack.spec import Spec
real_compilers = [{'compiler': {'environment': {}, 'extra_rpaths': [], 'flags': {}, 'implicit_rpaths': [], 'modules': [], 'operating_system': 'centos7', 'paths': {'cc': '/bin/gcc', 'cxx': '/bin/g++', 'f77': '/bin/gfortran', 'fc': '/bin/gfortran'}, 'spec': 'gcc@4.8.5', 'target': 'x86_64'}}]
with cfg.override('compilers:', real_compilers):
    s = Spec('gcc@5.5.0')
    s.concretize()
```

And what I get back is something like `gcc@5.5.0%gcc@5.5.0...`, which is not what I wanted.  So I feel it may be a bug that I cannot completely override the compiler config from within the environment.

### Steps to reproduce the issue

This is most easily and reliably reproduced in a docker container where the installed compilers are known.  

1. Start by putting the following `spack.yaml` into a directory (e.g. `/tmp/reproduce`):

```
spack:
  specs: [bzip2@1.0.8 os=centos7]
  mirrors: {mirror: 's3://spack-private/mirror'}
  gitlab-ci:
    mappings:
    - match: [os=centos7]
      runner-attributes:
        image:
          name: spack/centos7
          entrypoint: ['']
        tags: [docker]
        variables: {}
  compilers:
    - compiler:
        spec: gcc@5.5.0
        target: x86_64
        operating_system: centos7
        paths: { "cc": "", "cxx": "", "f77": "", "fc": "" }
        environment: {}
        extra_rpaths: []
        flags: {}
        modules: []
  repos: []
  upstreams: {}
  modules:
    enable: []
  packages: {}
  config: {}
```

2. In the same directory clone spack and make sure the latest `develop` branch is checked out.

3. Run the `spack/centos7` docker container, mounting that directory as `/work`:

```console
$ docker run --rm --entrypoint /bin/bash -v /tmp/reproduce:/work -ti spack/centos7
[root@ba3e3475c3c6 ~]# 
```

4. Change into the environment directory and source the newly cloned spack:

```console
[root@ba3e3475c3c6 ~]# cd /work/                
[root@ba3e3475c3c6 work]# source spack/share/spack/setup-env.sh
```

5. Now run a spack python session and try to manually override the compilers:

```console
[root@ba3e3475c3c6 work]# spack python
Spack version 0.12.1
Python 2.7.5, Linux x86_64
>>> import spack.config as cfg
>>> from spack.spec import Spec
>>> real_compilers = [{'compiler': {'environment': {}, 'extra_rpaths': [], 'flags': {}, 'implicit_rpaths': [], 'modules': [], 'operating_system': 'centos7', 'paths': {'cc': '/bin/gcc', 'cxx': '/bin/g++', 'f77': '/bin/gfortran', 'fc': '/bin/gfortran'}, 'spec': 'gcc@4.8.5', 'target': 'x86_64'}}]
>>> with cfg.override('compilers:', real_compilers):
...     s = Spec('gcc@5.5.0')
...     s.concretize() 
...     print(s)
... 
gcc@5.5.0%gcc@5.5.0...
>>>
```

Please let me know if more information is required.  Using the same docker container should allow faithful reproduction without worrying about different system configs.

**Comment by scottwittenburg** [comment]
@tgamblin Here's the issue we discussed earlier today.  Let me know if you have issues reproducing it, or if I can be of assistance in any way.  Thanks for taking a look!

**Comment by scheibelp** [comment]
Is there a reason you are doing this programatically vs. using the `::` syntax?  https://github.com/spack/spack/pull/11602 was merged recently and should allow this. If that does not work could you also let me know (or if that is not suitable for your use case could you also mention that here)?

Regardless of whether `::` works, the override syntax should also work, so this should stay open.

**Comment by scheibelp** [comment]
As another interim fix can you make use of the `-C` argument like `spack -C config-dir/` to "swap out" different configs?

**Comment by scottwittenburg** [comment]
In my use case, I'm iterating over the compilers in my configuration and checking each one to see if it has all empty `paths`, in which case I don't add it to my list of `real_compilers`.  I'm fine using the `::` syntax if that works, but I'm not sure I understand how to use it.  Does it just require writing the yaml file to disk?

**Comment by scottwittenburg** [comment]
I have successfully used the `-C` argument in the past, so I'll try it as a workaround in this case, thanks!

**Comment by scheibelp** [comment]
> Does it just require writing the yaml file to disk?

It does. The syntax should work as described in https://spack.readthedocs.io/en/latest/configuration.html#overriding-entire-sections. In this case you would want it to start out with `compilers::`.

Also, as I read the docs for `config.override` it says that it pushes another scope on top of the existing ones rather than that it replaces the existing configs with new configs. Therefore I don't think it was intended to do what you wanted (although the terminology of "override" vs. "complete override" is admittedly confusing). Was there a time when that was working in this process?

**Comment by scottwittenburg** [comment]
Thanks for the pointer, I'll do that.

> Was there a time when that was working in this process?

No, this is the first time I've tried it.

> I don't think it was intended to do what you wanted

Fair enough.  Maybe I'm trying to do something weird and need to rethink it.  I want everything from the environment except some of the compilers, which I can identify.  I'll mull it over.

**Comment by scheibelp** [comment]
OK in that case I think it's safe to close this once you get one of the above methods working since `config.override` was not intended for this.

**Comment by tgamblin** [comment]
@scheibelp @scottwittenburg: I think in this case there is a bit more to the issue, but I agree @scottwittenburg can make it work with the methods mentioned.

In this case, since the user constructs their environment for GitLab CI, I *think* it would be useful for *no* other config to be considered on the runners.  If you version your `spack.yaml` in a repo and you fire off GitLab runners with it, it seems like the runners should respect all the settings in the CI repo, and ignore any other configuration from their runtime environments.

I guess I could also imagine runners that want to make use of runner-specific `packages.yaml`, though, so maybe it's worth also allowing configs to be merged.

@scheibelp does `spack::` currently work to override everything but defaults?  That seems like it might be a good way to specify this -- `spack::` for "only what is in my environment" and `spack:` for "merge with host config".  It follows the existing convention but I am not sure that we handled `spack::`

**Comment by scheibelp** [comment]
> scheibelp does spack:: currently work to override everything but defaults?

No: that would take some extra updates to the config module. You can approximate this for now by overriding all the sections though.

**Comment by scottwittenburg** [comment]
@tgamblin This is the issue I mentioned earlier today, which is currently blocking my ability to move forward on bootstrapping compilers in the new pipeline workflow.

**Comment by scheibelp** [comment]
> can I "swap out" different configs within the context of that single spack invocation?

No (or at least that is definitely not supported well)

> Or maybe you were thinking I would invoke spack with a -C, maybe passing an initially empty directory, then as the spack ci rebuild ... command progresses, I can add/remove config files from that directory?

I would want to make it easier for you than that. If the config has to change part way through an invocation, and there's no way to separate out that one invocation into multiple separate invocations, then we'll have to chat about how to support that.

**Comment by scottwittenburg** [comment]
> If the config has to change part way through an invocation, and there's no way to separate out that one invocation into multiple separate invocations, then we'll have to chat about how to support that.

We could definitely separate it into several invocations.  Currently the build pipeline is just that, a bunch of separate invocations of spack commands from a few different bash scripts.  Now that I'm trying to convert those pipeline shell scripts to spack commands, this is one of the resulting obstacles.  So it might make the most sense to do as you suggest: back off a titch and split `spack ci rebuild` into a couple invocations.  I guess I especially agree if it seems like what I'm trying to do is pretty far outside the intended usage of the config system.

By the way, I created an issue (#13115) for discussion of how that `spack ci` cli api could look.  So I I just made some [comments](https://github.com/spack/spack/issues/13115#issuecomment-552038610) over there on how the split mentioned above could be made.

Thanks @scheibelp 😄and let me know your thoughts @tgamblin

**Comment by alalazo** [comment]
Closing since this seems outdated by "compiler as deps" in Spack v1.0 and "groups" in Spack v1.2
